src/libAtomVM/bif.c

```diff
@@ -343,6 +344,69 @@ term bif_erlang_map_get_2(Context *ctx, uint32_t fail_label, term arg1, term arg
     return term_get_map_value(arg2, pos);
 }
 
+#ifdef AVM_NO_SMP
+static int64_t get_unique_monotonic_integer(void)
+{
+    static int64_t unique = 0;
+
+    if (UNLIKELY(unique == INT64_MAX)) {
+        AVM_ABORT();
+    }
+    return unique++;
+}
+
+#elif ATOMIC_LLONG_LOCK_FREE == 2
+
+static int64_t get_unique_monotonic_integer(void)
+{
+    static int64_t ATOMIC unique = 0;
+
+    if (UNLIKELY(unique == INT64_MAX)) {
+        AVM_ABORT();
+    }
+    return unique++;
+}
+#else
+
+static int64_t get_unique_monotonic_integer(void)
+{
+    static int64_t unique = 0;
+    static SpinLock unique_spinlock;
+
+    if (UNLIKELY(unique == INT64_MAX)) {
+        AVM_ABORT();
+    }
+
+    smp_spinlock_lock(&unique_spinlock);
+    int64_t value = unique++;
+    smp_spinlock_unlock(&unique_spinlock);
+    return value;
+}
+
+#endif
+
+term bif_erlang_unique_integer_0(Context *ctx)
+{
+    int64_t value = get_unique_monotonic_integer();
+    return term_make_maybe_boxed_int64(value, &ctx->heap);
+}
+
+term bif_erlang_unique_integer_1(Context *ctx, uint32_t fail_label, term arg1)
+{
+    int proper = 0;
+    if (UNLIKELY(!term_is_list(arg1))) {
+        RAISE_ERROR_BIF(fail_label, BADARG_ATOM);
+    }
+    size_t _len = term_list_length(arg1, &proper);
+    UNUSED(_len);
+    if (UNLIKELY(!proper)) {
+        RAISE_ERROR_BIF(fail_label, BADARG_ATOM);
+    }
```

Comment:
I kept this check to prevent incorrect arg passing. Can be removed deemed unnecessary.